### PR TITLE
SOFIE + Keras TF for FlagBkgHits using Stereo Hits

### PR DESCRIPTION
**Pull request by myucel-fnal**

### CODE REVIEW

**Comment by kutschke** [comment]
@FNALbuild run build test

**Comment by brownd1978** [comment]
@FNALbuild run build test

**Comment by brownd1978** [comment]
@FNALbuild run build test

File: RecoDataProducts/src/ComboHit.cc [hunk 1]
```diff
@@ -146,7 +146,7 @@ namespace mu2e {
           for(size_t ich=0;ich < size();++ich){
             ComboHit const& ch = (*this)[ich];
             for(auto iph : ch.indexArray())
-              shids[ich].insert(shids[ich].end(),pshids[iph].begin(),pshids[iph].end());
+              if(pshids.size() == 8) shids[ich].insert(shids[ich].end(),pshids[iph].begin(),pshids[iph].end());
```
Comment (brownd1978) [inline RecoDataProducts/src/ComboHit.cc:149]: This effectively turns off filling these indices, as the # of hits is never(?) 8.   I suspect the real problem is how these indices are being interpreted in FlagBkgHits.  I suggest to revert this change and create an issue, specific to stereo hits.

File: TrkDiag/fcl/prolog.fcl [hunk 2]
```diff
@@ -8,7 +8,7 @@ BD : {
   StrawDigiMCCollection : "compressDigiMCs"
   ComboHitCollection : "makePH"
   StrawHitFlagCollection : "FlagBkgHits:ComboHits"
-  BkgQualCollection:"FlagBkgHits"
+  # BkgQualCollection:"FlagBkgHits"
```
Comment (brownd1978) [inline TrkDiag/fcl/prolog.fcl:11]: please remove the commented line, git history will keep track of this change

File: TrkDiag/src/BkgDiag_module.cc [hunk 3]
```diff
@@ -313,8 +303,16 @@ namespace mu2e
       double sqrSumDeltaTime(0.);
       double sqrSumDeltaX(0.);
       double sqrSumDeltaY(0.);
+      float zmin(0.);
+      float zmax(0.);
+      std::vector<float> hz;
+      std::vector<int> hp;
       for(auto const& ich : cluster.hits()){
         ComboHit const& ch = _chcol->at(ich);
+        if(ch.pos().Z() < zmin) zmin = ch.pos().Z();
```
Comment (brownd1978) [inline TrkDiag/src/BkgDiag_module.cc:312]: zmin and zmax seem irrelevant now you have a sorted vector to fill _zmin and _zmax

**Review by brownd1978 — CHANGES_REQUESTED** [review]
Just a few small things to fix

**Comment by brownd1978** [comment]
@FNALbuild run build test

**Review by brownd1978 — COMMENTED** [review]
Another small issue

File: TrkHitReco/src/FlagBkgHits_module.cc [hunk 4]
```diff
@@ -255,86 +251,72 @@ namespace mu2e
...
-      fillMVA(cqual);
 
-      StrawHitFlag flag(StrawHitFlag::bkgclust);
-      if (cqual.MVAOutput() > bkgMVAcut_)
+      unsigned nactive, nstereo;
+      countHits(cluster, nactive, nstereo,chcol);
+      unsigned nhits = nstereo > 0 ? nstereo : nactive;
+      if (nhits < minnhits_) return;
+      std::vector<float> kerasvars(11,0.0);
+      kerasvars.at(0) = sqrtf(cluster.pos().perp2());
```
[the remark this answers is not in the document]
Comment (myucel-fnal) [reply TrkHitReco/src/FlagBkgHits_module.cc]: Did that, plus changed std::vector to std::array. Due to std::array being created on stack it is supposed to be faster.

File: TrkHitReco/src/FlagBkgHits_module.cc [hunk 5]
```diff
@@ -255,86 +256,72 @@ namespace mu2e
...
-      fillMVA(cqual);
 
-      StrawHitFlag flag(StrawHitFlag::bkgclust);
-      if (cqual.MVAOutput() > bkgMVAcut_)
+      unsigned nactive, nstereo;
+      countHits(cluster, nactive, nstereo,chcol);
+      unsigned nhits = nstereo > 0 ? nstereo : nactive;
+      if (nhits < minnhits_) return;
+      std::array<float,11> kerasvars;
+      kerasvars[0] = sqrtf(cluster.pos().perp2());
```
Comment (brownd1978) [inline TrkHitReco/src/FlagBkgHits_module.cc:265]: This is the same as cluster.pos.Rho()

Comment (myucel-fnal) [reply TrkHitReco/src/FlagBkgHits_module.cc:265]: yes, do you want me to change it?

**Comment by kutschke** [comment]
@FNALbuild run build test

**Pull request merged**